Add new `treefile-apply` experimental command

jlebon opened the pull request:
We are working towards rebasing FCOS and RHCOS on top of fedora-bootc
and rhel-bootc. As one can imagine, we are quite heavily invested in
treefiles currently, so being able to keep using what we already have as
we transition into becoming a layered build is valuable.

Dockerfiles are great of course as the common language used to build
container images. But it's not great for implementing complex logic.

Add a new `rpm-ostree experimental compose treefile-apply` command
which takes a treefile, but applies its directives _live_ in the running
environment.

So in this model, the business logic in the Dockerfile is kept to a
minimum and all the heavy lifting happens via
`RUN rpm-ostree compose treefile-apply manifest.yaml`.

A very small subset of the treefile spec is actually supported. The goal
is _not_ to reimplement everything, but only what is needed and useful
in a derivation context.

This code actually started as a Python script in openshift/os, currently
used to build the OCP node image:

https://github.com/openshift/os/blob/master/scripts/apply-manifest

So in fact, the node image (which in turn is a layered build on top of
RHCOS) will also be using this code.

cgwalters commented on rust/src/treefile.rs:
```diff
@@ -4668,3 +4670,82 @@ pub(crate) fn treefile_delete_client_etc() -> CxxResult<u32> {
+    cmd.arg("--noplugins");
+    cmd.arg("-y");
+
+    match treefile.parsed.documentation {
```
It'd seem to me to be quite unusual to override the documentation flag in a *derived* build; isn't a feature here to inherit the base image's `/etc/dnf.conf`?

Maybe just error out if the option is set?

jlebon replied:
If a user wants to inherit, they can leave `documentation` unspecified. If they want documentation for their layered packages even if the base image didn't have them, I feel like we should allow that? There's nothing technically that prevents it.

cgwalters replied:
> If they want documentation for their layered packages even if the base image didn't have them, I feel like we should allow that? There's nothing technically that prevents it.

I agree technically it's possible but I feel like it's at best a really unusual situation we could not care about or try to test - the 95% case I suspect is folks who would want to take a base image for which documentation is turned off and ensure that documentation is turned on for the base image and all the layered stuff too.

That'd be doable, in theory if we reinstall those base packages with tsflags=docs (i.e. !nodocs) we should end up in mostly the same state. Except of course for packages where we do file remapping like selinux.

cgwalters replied:
But if you want to keep this as is, that's also OK by me and if so feel free to just mark this conversation as resolved!

jlebon replied:
WFM. We can delay until there's a concrete ask for it.

jlebon replied:
Instead of making it a hard error, I switched it to be a warning to make it easier to use with treefiles actively used for base composes as well.

cgwalters replied:
Fine by me but wouldn't it be better to split up those two treefiles though?

cgwalters commented on rust/src/treefile.rs:
```diff
@@ -4668,3 +4670,82 @@ pub(crate) fn treefile_delete_client_etc() -> CxxResult<u32> {
+        None => {} // implicitly leave environment default if unset
+    }
+
+    // we support an API in which yum repos can be injected in the container
```
What build process does that? Is this something in cosa or prow specifically? It makes sense, but would also make sense to argue to genereralize I think

jlebon replied:
Made it up myself as part of the layered node image work: https://github.com/openshift/os/blob/89030b9b283d4ef6db37f8d761bb8996557fab3e/Containerfile#L6-L8. But really, I should be using the secrets API instead for this. Will rework this.

jlebon replied:
Opened https://github.com/openshift/os/pull/1740 and dropped this!

cgwalters commented on rust/src/treefile.rs:
```diff
@@ -4668,3 +4670,82 @@ pub(crate) fn treefile_delete_client_etc() -> CxxResult<u32> {
+
+    let status = cmd.status().context("collecting dnf status")?;
+    if !status.success() {
+        bail!("Failed to run skopeo copy: {status:?}");
```
not skopeo copy

cgwalters commented on rust/src/treefile.rs:
```diff
@@ -4668,3 +4670,82 @@ pub(crate) fn treefile_delete_client_etc() -> CxxResult<u32> {
+    }
+
+    let status = cmd.status().context("collecting dnf status")?;
+    if !status.success() {
```
I think we could use the handy `.run()` helper from bootc-utils

cgwalters commented:
I think we could use a test case for this, should be pretty easy right? `test-container.sh` could work, or just a new Dockerfile, etc.

Anything else keeping this in draft?

jlebon commented:
Added test, rebased and moved out of draft! There's definitely more to add here, but it's enough [for some success](https://github.com/coreos/fedora-coreos-config/pull/3348#issuecomment-2643694571).

cgwalters reviewed (approved):
LGTM!

cgwalters commented:
Looks like CI failed actually with 
```
 + rpm-ostree experimental compose treefile-apply /tmp/treefile.yaml
error: No such file or directory (os error 2)
error: locking base packages with dnf: Failed to run dnf versionlock: ExitStatus(unix_wait_status(256))
```

jlebon commented:
CI fix in https://github.com/coreos/rpm-ostree/pull/5286. I guess let's make it required after that merges?
